File: command/configure.py
from typing import NoReturn

import logging
import subprocess

from command import boot
from globals import (
    BIN_GENERATE_DRIVER_PATH,
    SAVE_DRIVER_FOLDER_PATH,
    find_grayscale_camera,
    ExitCode,
)


def configure(
    device: str | None, manual: bool, emitters: int, neg_answer_limit: int
) -> NoReturn:
    """Find a driver for the infrared camera and exit.

    Args:
        device (str | None): path to the infrared camera, None for automatic detection
        manual (bool): true for enabling the manual configuration
        emitters (int): number of emitters on the device
        neg_answer_limit (int): number of negative answer before the pattern is skiped. Use -1 for unlimited
    """
    logging.info(
        "Stand in front of and close to the camera and make sure the room is well lit."
    )
    logging.info("Ensure to not use the camera during the execution.")
    logging.info("Warning to do not kill the process !")

    if device is None:
        device = find_grayscale_camera()

    logging.info(f"Configuring the camera: {device}.")
    log_level = int(logging.getLogger().level == logging.DEBUG)
    exit_code = subprocess.call(
        [
            BIN_GENERATE_DRIVER_PATH,
            device,
            str(emitters),
            str(neg_answer_limit),
            SAVE_DRIVER_FOLDER_PATH,
            str(log_level),
            str(int(manual)),
        ]
    )

    if exit_code != ExitCode.SUCCESS:
        logging.error("The configuration has failed.")
        logging.error("Please retry in manual mode by adding the '-m' option.")
        logging.info("Do not hesitate to visit the GitHub !")
        logging.info(
            "https://github.com/EmixamPP/linux-enable-ir-emitter/blob/master/docs/README.md"
        )
    else:
        logging.info("The driver has been successfully generated.")
        boot("enable")

    exit(exit_code)

File: command/configure.cpp
#include <fstream>
#include <memory>
using namespace std;

#include <spdlog/spdlog.h>

#include "camera/autocamera.hpp"
#include "camera/camerainstruction.hpp"
#include "commands.hpp"
#include "configuration/finder.hpp"
#include "configuration/scanner.hpp"

ExitCode configure(const optional<string> &device, int width, int height, bool manual,
                   unsigned emitters, unsigned neg_answer_limit, bool no_gui) {
  spdlog::debug("Executing configure command.");
  spdlog::info("Stand in front of and close to the camera and make sure the room is well lit.");
  spdlog::info("Ensure to not use the camera during the execution.");

  bool success = false;
  try {
    shared_ptr<Camera> camera;
    if (manual)
      camera = CreateCamera<Camera>(device, width, height, no_gui);
    else
      camera = CreateCamera<AutoCamera>(device, width, height, no_gui);

    spdlog::info("Configuring the camera {}.", camera->device());

    auto instructions = Configuration::LoadInit(camera->device());
    if (!instructions) {
      spdlog::debug("No previous configuration found.");
      Scanner scanner(camera);
      instructions = scanner.scan();
      Configuration::SaveInit(camera->device(), instructions.value());
    } else
      spdlog::debug("Previous configuration found.");

    Finder finder(camera, emitters, neg_answer_limit);

    if (camera->Camera::is_emitter_working()) {
      spdlog::error("The emitter is already working, skipping the configuration.");
      return ExitCode::FAILURE;
    }

    success = finder.find(instructions.value());
    success = success && Configuration::Save(camera->device(), instructions.value());
  } catch (const CameraException &e) {
    spdlog::critical(e.what());
    exit(ExitCode::FILE_DESCRIPTOR_ERROR);
  }

  if (!success) {
    spdlog::error("The configuration failed.");
    spdlog::info("Please retry in manual mode by adding the '-m' option.");
    spdlog::info("Do not hesitate to visit the GitHub!");
    spdlog::info("https://github.com/EmixamPP/linux-enable-ir-emitter/blob/master/docs/README.md");
    return ExitCode::FAILURE;
  }

  spdlog::info("The infrared camera has been successfully configured.");
  return ExitCode::SUCCESS;
}
